lib/Dialect/RTL/RTLFolds.cpp

```diff
@@ -732,3 +732,219 @@ void WireOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
   };
   results.insert<DropDeadConnect>(context);
 }
+
+//===----------------------------------------------------------------------===//
+// MuxOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult MuxOp::fold(ArrayRef<Attribute> constants) {
+  // mux (c, b, b) -> b
+  if (trueValue() == falseValue())
+    return trueValue();
+
+  // mux(0, a, b) -> b
+  // mux(1, a, b) -> a
+  if (auto pred = constants[0].dyn_cast_or_null<IntegerAttr>()) {
+    if (pred.getValue().isNullValue())
+      return falseValue();
+    return trueValue();
+  }
+
+  return {};
+}
+
+//===----------------------------------------------------------------------===//
+// ICmpOp
+//===----------------------------------------------------------------------===//
+
+// Calculate the result of a comparison when the LHS and RHS are both
+// constants.
+static bool applyCmpPredicate(ICmpPredicate predicate, const APInt &lhs,
+                              const APInt &rhs) {
+  switch (predicate) {
+  case ICmpPredicate::eq:
+    return lhs.eq(rhs);
+  case ICmpPredicate::ne:
+    return lhs.ne(rhs);
+  case ICmpPredicate::slt:
+    return lhs.slt(rhs);
+  case ICmpPredicate::sle:
+    return lhs.sle(rhs);
+  case ICmpPredicate::sgt:
+    return lhs.sgt(rhs);
+  case ICmpPredicate::sge:
+    return lhs.sge(rhs);
+  case ICmpPredicate::ult:
+    return lhs.ult(rhs);
+  case ICmpPredicate::ule:
+    return lhs.ule(rhs);
+  case ICmpPredicate::ugt:
+    return lhs.ugt(rhs);
+  case ICmpPredicate::uge:
+    return lhs.uge(rhs);
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+// Returns the result of applying the predicate when the LHS and RHS are the
+// exact same value.
+static bool applyCmpPredicateToEqualOperands(ICmpPredicate predicate) {
+  switch (predicate) {
+  case ICmpPredicate::eq:
+  case ICmpPredicate::sle:
+  case ICmpPredicate::sge:
+  case ICmpPredicate::ule:
+  case ICmpPredicate::uge:
+    return true;
+  case ICmpPredicate::ne:
+  case ICmpPredicate::slt:
+  case ICmpPredicate::sgt:
+  case ICmpPredicate::ult:
+  case ICmpPredicate::ugt:
+    return false;
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
+
+  // gt a, a -> false
+  // gte a, a -> true
+  if (lhs() == rhs()) {
+    auto val = applyCmpPredicateToEqualOperands(predicate());
+    return IntegerAttr::get(getType(), val);
+  }
+
+  auto lhs = constants[0].dyn_cast_or_null<IntegerAttr>();
+  auto rhs = constants[1].dyn_cast_or_null<IntegerAttr>();
```

Comment:
Nit-pick, but I'd microoptimize this like this to avoid the RHS check when the LHS is non-constant.  Between the two of them the LHS is more likely to be a non-constant:

```
if (auto lhs = constants[0].dyn_cast_or_null<IntegerAttr>()) {
  if (auto rhs = constants[1].dyn_cast_or_null<IntegerAttr>()) {
...
```